tls/s2n_tls13_secrets.c

```diff
@@ -636,6 +658,15 @@ S2N_RESULT s2n_tls13_secrets_update(struct s2n_connection *conn)
                     S2N_CLIENT, &CONN_SECRET(conn, client_app_secret)));
             RESULT_GUARD(s2n_tls13_derive_secret(conn, S2N_MASTER_SECRET,
                     S2N_SERVER, &CONN_SECRET(conn, server_app_secret)));
+            RESULT_GUARD(s2n_derive_exporter_master_secret(conn,
+                        &CONN_SECRET(conn, exporter_master_secret)));
+
+            if (conn->secret_cb && (s2n_connection_is_quic_enabled(conn) || s2n_in_unit_test())) {
+                RESULT_GUARD_POSIX(conn->secret_cb(conn->secret_cb_context, conn, S2N_EXPORTER_SECRET,
+                            CONN_SECRET(conn, exporter_master_secret).data, CONN_SECRET(conn, exporter_master_secret).size));
+            }
+            s2n_result_ignore(s2n_key_log_tls13_secret(conn, &CONN_SECRET(conn, exporter_master_secret), S2N_EXPORTER_SECRET));
```

Comment:
I'm not a huge fan of this - duplicates the quic & keylog function calls from the s2n_trigger_secret_callbacks function. That takes a s2n_extract_secret_type_t which doesn't feel like it should be extended to have a EXPORTER_SECRET in it, since the exporter *isn't* produced via HKDF-Extract... but maybe the naming scheme should instead be updated?

I haven't investigated deeply on the reasons for the separate s2n_extract_secret_type_t and s2n_secret_type_t enums and whether there's some unification or other changes possible that would fit well.

Reply:
Minimally, this isn't the right place for this logic. It would belong in s2n_derive_exporter_master_secret. The current call to s2n_trigger_secret_callbacks is made in s2n_tls13_derive_secret, not here. s2n_tls13_secrets_update is just supposed to trigger the various secret updates, and it needs to be kept very simple.

I'm not sure about adding EXPORTER_SECRET to s2n_extract_secret_type_t. It wouldn't be allowed other places that enum is used, like [when setting keys](https://github.com/aws/s2n-tls/blob/ab7bfce9b78109c26600d1bb63c555643be1c345/tls/s2n_tls13_key_schedule.c#L50). A better idea might be to factor out the triggering logic from the type translation logic.

If you don't immediately need the callbacks / logging, you could also not enable it as part of this PR and just open an issue to add it later.

Reply:
@camshaft I think we discussed possibly needing the callback to get at the exported material in s2n-quic ([here](https://github.com/aws/s2n-quic/blob/main/quic/s2n-quic-tls/src/callback.rs#L160)). I think we can maybe make it work without that, but it'll be trickier. If we expect to need it or benefit from it, I'm inclined to retain the callbacks in this PR.

> It would belong in s2n_derive_exporter_master_secret. The current call to s2n_trigger_secret_callbacks is made in s2n_tls13_derive_secret, not here. 

I'll make this change in a bit.

Reply:
I don't _think_ we need an explicit callback for it on the QUIC side... It would probably be easier to assume that once we get application keys we are able to call the connection's exporter function.

Reply:
And it can always be a follow-up PR. Especially if it involves some non-trivial refactoring, you should probably be trying to keep this PR small.